File: backend/components/metiers/administration_publique/utils/exporter/core/exporter.test.js
// exporter.test.js – Tests unitaires JS pour exporter Threed
const exporter = require('./exporter');
const fs = require('fs');
const path = require('path');

describe('Exporter Threed', () => {
  const testData = [
    { id: 1, user: 'Alice', value: 42 },
    { id: 2, user: 'Bob', value: 99 }
  ];
  const jsonPath = path.join(__dirname, 'test_export.json');
  const csvPath = path.join(__dirname, 'test_export.csv');

  afterAll(() => {
    if (fs.existsSync(jsonPath)) fs.unlinkSync(jsonPath);
    if (fs.existsSync(csvPath)) fs.unlinkSync(csvPath);
  });

  it('doit exporter en JSON', () => {
    exporter.exportToJSON(testData, jsonPath);
    const content = fs.readFileSync(jsonPath, 'utf-8');
    expect(content).toMatch('Alice');
    expect(content).toMatch('Bob');
  });

  it('doit exporter en CSV', () => {
    exporter.exportToCSV(testData, csvPath);
    const content = fs.readFileSync(csvPath, 'utf-8');
    expect(content).toMatch('Alice');
    expect(content).toMatch('Bob');
    expect(content).toMatch('id,user,value');
  });

  it('doit anonymiser les données RGPD', () => {
    const anonymized = exporter.anonymize(testData);
    expect(anonymized[0].user).toBe('anonyme');
    expect(anonymized[1].user).toBe('anonyme');
  });

  it('doit gérer le cas d’un tableau vide', () => {
    const result = exporter.exportToCSV([], csvPath);
    expect(result).toBe(csvPath);
  });
});

File: backend/components/metiers/administration_publique/utils/exporter/core/exporter.test.py
"""
exporter.test.py – Tests unitaires Python pour exporter Threed
"""
import os
import json
import csv
from .exporter import export_to_json, export_to_csv, anonymize

def test_export_to_json(tmp_path):
    data = [{'id': 1, 'user': 'Alice'}, {'id': 2, 'user': 'Bob'}]
    file_path = tmp_path / 'test_export.json'
    export_to_json(data, file_path)
    with open(file_path, encoding='utf-8') as f:
        content = f.read()
        assert 'Alice' in content and 'Bob' in content

def test_export_to_csv(tmp_path):
    data = [{'id': 1, 'user': 'Alice'}, {'id': 2, 'user': 'Bob'}]
    file_path = tmp_path / 'test_export.csv'
    export_to_csv(data, file_path)
    with open(file_path, encoding='utf-8') as f:
        content = f.read()
        assert 'Alice' in content and 'Bob' in content and 'id,user' in content

def test_anonymize():
    data = [{'id': 1, 'user': 'Alice'}, {'id': 2, 'user': 'Bob'}]
    anonymized = anonymize(data)
    assert all(row['user'] == 'anonyme' for row in anonymized)

def test_export_empty():
    file_path = 'empty.csv'
    result = export_to_csv([], file_path)
    assert result == file_path
    if os.path.exists(file_path):
        os.remove(file_path)

def test_export_large_data(tmp_path):
    data = [{'id': i, 'user': f'User{i}'} for i in range(1000)]
    file_path_json = tmp_path / 'test_export_large.json'
    file_path_csv = tmp_path / 'test_export_large.csv'
    export_to_json(data, file_path_json)
    export_to_csv(data, file_path_csv)
    with open(file_path_json, encoding='utf-8') as f:
        content = f.read()
        assert all(f'User{i}' in content for i in range(1000))
    with open(file_path_csv, encoding='utf-8') as f:
        content = f.read()
        assert all(f'User{i}' in content for i in range(1000)) and 'id,user' in content

def test_anonymize_edge_case():
    data = [{'id': 1, 'user': None}, {'id': 2, 'user': ''}]
    anonymized = anonymize(data)
    assert all(row['user'] == 'anonyme' for row in anonymized)

def test_export_special_characters(tmp_path):
    data = [{'id': 1, 'user': 'Al!ce'}, {'id': 2, 'user': 'B@b'}]
    file_path_json = tmp_path / 'test_export_special.json'
    file_path_csv = tmp_path / 'test_export_special.csv'
    export_to_json(data, file_path_json)
    export_to_csv(data, file_path_csv)
    with open(file_path_json, encoding='utf-8') as f:
        content = f.read()
        assert 'Al!ce' in content and 'B@b' in content
    with open(file_path_csv, encoding='utf-8') as f:
        content = f.read()
        assert 'Al!ce' in content and 'B@b' in content and 'id,user' in content

def test_export_invalid_data(tmp_path):
    data = "This is not a valid data format"
    file_path = tmp_path / 'test_export_invalid.json'
    try:
        export_to_json(data, file_path)
        exported = True
    except Exception:
        exported = False
    assert not exported

def test_anonymize_performance(benchmark):
    data = [{'id': i, 'user': 'User{i}'} for i in range(10000)]
    benchmark(anonymize, data)
